mlflow/pipelines/cards/pandas_renderer.py

```diff
@@ -0,0 +1,244 @@
+    :return: A DatasetFeatureStatistics proto.
+    """
+    fs_proto = facet_feature_statistics_pb2.FeatureNameStatistics
+    feature_stats = facet_feature_statistics_pb2.DatasetFeatureStatistics()
+    pandas_describe = df.describe(datetime_is_numeric=True, include="all")
+    feature_stats.num_examples = len(df)
+    quantiles_to_get = [x * 10 / 100 for x in range(10 + 1)]
+    quantiles = df.quantile(quantiles_to_get)
+
+    for key in df:
+        pandas_describe_key = pandas_describe[key]
+        current_column_value = df[key]
+        feat = feature_stats.features.add(
+            type=DtypeToType(current_column_value.dtype), name=key.encode("utf-8")
+        )
+        if feat.type in (fs_proto.INT, fs_proto.FLOAT):
+            feat_stats = feat.num_stats
+
+            converter = DtypeToNumberConverter(current_column_value.dtype)
+            if converter:
+                date_time_converted = converter(current_column_value)
+                current_column_value = pd.DataFrame(date_time_converted)[0]
+                pandas_describe_key = current_column_value.describe(
+                    datetime_is_numeric=True, include="all"
+                )
+                quantiles[key] = current_column_value.quantile(quantiles_to_get)
+
+            default_value = 0
+            feat_stats.std_dev = pandas_describe_key.get("std", default_value)
+            feat_stats.mean = pandas_describe_key.get("mean", default_value)
+            feat_stats.min = pandas_describe_key.get("min", default_value)
+            feat_stats.max = pandas_describe_key.get("max", default_value)
+            feat_stats.median = current_column_value.median()
+            feat_stats.num_zeros = (current_column_value == 0).sum()
+            feat_stats.common_stats.CopyFrom(compute_common_stats(current_column_value))
+
+            if key in quantiles:
+                equal_width_hist = histogram_generator.generate_equal_width_histogram(
+                    quantiles=quantiles[key].to_numpy(),
+                    num_buckets=10,
+                    total_freq=feat_stats.common_stats.num_non_missing,
+                )
+                if equal_width_hist:
+                    feat_stats.histograms.append(equal_width_hist)
+                equal_height_hist = histogram_generator.generate_equal_height_histogram(
+                    quantiles=quantiles[key].to_numpy(), num_buckets=10
+                )
+                if equal_height_hist:
+                    feat_stats.histograms.append(equal_height_hist)
+        elif feat.type == fs_proto.STRING:
+            feat_stats = feat.string_stats
+            strs = []
+            compute_unique_str = current_column_value.dropna()
+            for item in compute_unique_str:
+                strs.append(
+                    item
+                    if hasattr(item, "__len__")
+                    else item.encode("utf-8")
+                    if hasattr(item, "encode")
+                    else str(item)
```

Comment:
Will these 2 lines ever be called?

Reply:
Got this code from here: https://github.com/PAIR-code/facets/blob/4742b8b93c2dacf22fc8ace2cee42dd06382c48e/facets_overview/facets_overview/base_generic_feature_statistics_generator.py#L263-L265
I assumed that this might be required for an edge case when the strings are not encoded. So left this code path here. 

Do you think we need to support this usecase? I am happy to remove it

Reply:
No I'm just questioning the usage of 2 consecutive if-else clauses. I doubt the 2nd one will ever get called. In what circumstances will it happen? It'd be great to demonstrate with an example.